### PR TITLE
Provides instructions on parse error caused by code scaffolding in learnr tutorial

### LINKED ISSUE #134: Default feedback for "fill in the blanks" blanks left in exercise
Many tutorial authors use exercises where they have a "fill in the blanks" type scaffolding for learners to start with, e.g.

```r
ggplot(data = ___, aes(x = mpg, y = ___) +
  ___
```

Usually, these are 3 underscores, but different authors might use different placeholders for the blanks.

It would be nice if gradethis had a default message if learner left any `___` in their submission, like _"Please replace all blanks with valid code."_ It's a bit of a pain to write code checking for existence of blanks and makes the `-check` chunk repetitive throughout the document, and this would help. I think tutorial authors would be fine if the package required a certain type of representation for blanks (e.g., 3 underscores) but it might also be nice if the authors could set a gradethis option within a tutorial to indicate the text string they're using to indicate blanks.

### PR DESCRIPTION
**Pull request by gadenbuie**
Fixes #134

Currently, the code scaffolding commonly used in exercises, e.g. `____(mtcars, ____)`, causes a parse error, resulting in an error message like the one below:

image

This PR adds logic to the default user code error handler in `grade_learnr()` to detect three or more `_` in the user code and produce an informative error message with instructions.

image

Because the logic adds a bit of complexity to the error handler, I moved the default handler into a new function: `grade_learnr_parse_error()`.

### CODE REVIEW

**Comment by gadenbuie** [comment]
Addresses #134

**Comment by schloerke** [comment]
Thank you!

File: R/grade_learnr.R [hunk 1]
```diff
@@ -186,6 +171,37 @@ grade_learnr_ <- function(label = NULL,
...
+  } else {
+    paste(
+      "Uh oh, the R code produced a syntax error:",
+      conditionMessage(parse_error),
+      "\nCheck that you have closed every \", ', (, and { ",
+      "with a matching \", ', ), and }. Also look for missing ",
+      "commas. R cannot determine how to turn your text into ",
+      "a complete command."
+    )
+  }
```
Comment (schloerke) [inline R/grade_learnr.R:201]: Cosmetic code formatting

```suggestion
  msg <- 
    if (n_blanks > 0) {
      paste0(
        "The exercise contains ", 
        if (n_blanks == 1L) {
          "1 blank"
        } else {
          paste(n_blanks, "blanks")
        },
        ". Please replace the '____' with valid R code."
      )
    } else {
      paste(
        "Uh oh, the R code produced a syntax error:",
        conditionMessage(parse_error),
        "\nCheck that you have closed every \", ', (, and { ",
        "with a matching \", ', ), and }. Also look for missing ",
        "commas. R cannot determine how to turn your text into ",
        "a complete command."
      )
    }
```

**Review by schloerke — APPROVED** [review]

**Comment by gadenbuie** [comment]
@schloerke @garrettgman this is ready for re-review.

In short, this PR adds `grade_parse_error()` an internal function that demonstrates how parse errors can be handled via the `exercise.parse.error` option. The option accepts a function with a single argument. 

The only argument to the parse error function is the gradethis check environment, typically called `check_obj`, which provides access to everything `gradethis` knows about the exercise. Most relevant are `.user_code` and `.error` which provide a character string of the user-entered code and the error condition thrown from the parse failure. (These are elements in `check_obj` and are accessed via `check_obj$.user_code`.)

Finally, the parse error function should return a `fail()` (or a `pass() if you want) with an appropriate message.

File: tests/testthat/test_zzzz-grade_learnr.R [hunk 2]
```diff
@@ -1,86 +1,73 @@
...
+    is_correct = FALSE,
+    msg = "contains 3 blanks"
+  )
+  
+  # Can customize the feedback through an exercise.parse.error function
+  parse_error_func <- function(x) {
+    graded(
+      correct = FALSE,
+      message = paste(
+        "The user code of", shQuote(x$.user_code),
```
Comment (schloerke) [inline tests/testthat/test_zzzz-grade_learnr.R:60]: Should use single quotes with no _fancy_ quotes.

```suggestion
        "The user code of", sQuote(x$.user_code, q = FALSE),
```

File: R/grade_learnr.R [hunk 3]
```diff
@@ -202,6 +173,14 @@ grade_learnr_ <- function(
     }
   )
 
+  tryCatch(
+    parse(text = user_code %||% ""),
+    error = function(e) {
+      check_obj_envir$.error <- e
+      to_check_fn <<- getOption("exercise.parse.error", grade_parse_error)
```
Comment (schloerke) [inline R/grade_learnr.R:180]:
```suggestion
      # Add the error object to the checking object
      check_obj_envir$.error <- e
      # Overwrite `to_check_fn` to validate the parse error function accepts `check_obj_envir`
      to_check_fn <<- getOption("exercise.parse.error", grade_parse_error)
```

**Review by schloerke — APPROVED** [review]
LGTM given small changes!

File: tests/testthat/test_zzzz-grade_learnr.R (hunk 2 above)
Comment (schloerke) [inline tests/testthat/test_zzzz-grade_learnr.R]: Booo. `sQuote(q)` was add in R 3.5. 

```suggestion
        "The user code of", paste0("'", x$.user_code, "'"),
```

Comment (gadenbuie) [reply tests/testthat/test_zzzz-grade_learnr.R]: Oops! Helpfully we only use it in a test 😌

File: tests/testthat/test_zzzz-grade_learnr.R [hunk 4]
```diff
@@ -1,86 +1,72 @@
...
-  expect_correct(
-    test_grade_learnr(
-      user_code = "4",
-      check_code = "grade_result(
+  expect_grade_learnr(
+    user_code = "4",
+    check_code = "grade_result(
           pass_if(~ .result == 4, 'yes. you did it.'),
           correct = 'this other correct message.',
           glue_correct = '{ .message } { .correct }'
```
Comment (schloerke) [inline tests/testthat/test_zzzz-grade_learnr.R:9]:
```suggestion
      pass_if(~ .result == 4, 'yes. you did it.'),
      correct = 'this other correct message.',
      glue_correct = '{ .message } { .correct }'
```

File: tests/testthat/test_zzzz-grade_learnr.R [hunk 5]
```diff
@@ -1,86 +1,72 @@
...
-        )"
-    ),
-    "yes. you did it. this other correct message."
+        )",
+  is_correct = TRUE,
+  msg = "yes. you did it. this other correct message."
   )
```
Comment (schloerke) [inline tests/testthat/test_zzzz-grade_learnr.R:13]:
```suggestion
    )",
    is_correct = TRUE,
    msg = "yes. you did it. this other correct message."
  )
```

File: tests/testthat/test_zzzz-grade_learnr.R (hunk 4 above)
Comment (schloerke) [inline tests/testthat/test_zzzz-grade_learnr.R:9]:
```suggestion
      pass_if(~ .result == 4, 'yes. you did it.'),
      correct = 'this other correct message.',
      glue_correct = '{ .message } { .correct }'
```

**Review by schloerke — APPROVED** [review]

**Pull request merged**